futures-02.md

```diff
@@ -0,0 +1,396 @@
+  and their dependencies.
+- `futures-io` which includes the updated `AsyncRead` and `AsyncWrite` traits.
+- `futures-sink` for the `Sink` trait which is used for sending values
+  asynchronously.
+- `futures-executor` for the `Executor` trait which allows spawning futures
+  to run asynchronously.
+- `futures-channel` for channels and other synchronization primitives.
+- `futures-util` which offers a number of convenient combinator functions,
+  some of which are included in `FutureExt` and `StreamExt` extension traits.
+
+By minimizing the number of `futures` crates in the public API of your library,
+you will be able to receive more frequent updates without causing breaking
+changes for your users.
+
+Application developers can either depend on individual crates or on the
+`futures` crate, which provides convenience reexports of all the `futures-xxx`
+crates. Expect that this crate will publish more frequent semver-breaking
+changes, but that updating will not normally result in significant or
+widespread breakage.
+
+This change comes alongside he release of the new `tokio` crate.
+These crates offer greatly easier-to-use and easier-to-evolve APIs.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Below you will find a partial list of some of the breaking changes we hope to
+include in the new `futures` 0.2 release. Note that this list is not exact,
+and does not aim to be so: this list is merely intended to motivate and
+explain some of the desired changes to the `futures` create. Please leave
+comments about specific issues on [the `futures` repo](https://github.com/alexcrichton/futures-rs),
+as detailed discussion could easily overwhelm the RFC thread.
+
+## Explicit Task Context and Wakeup Handles
+
+The `Future`, `Stream`, `AsyncRead`, and `AsyncWrite` traits have been modified
+to take explicit `task::Context` arguments. `task::Context` is a type which
+provides access to
+[task-locals](https://docs.rs/futures/*/futures/macro.task_local.html) and a `Waker`:
+
+```rust
+mod task {
+
+  /// The context for the currently running task.
+  pub struct Context {
+
+      /// Returns a `Waker` which can be used to wake up the current task.
+      pub fn waker(&self) -> Waker { ... }
+
+      /// Runs the provided closure with the task-local pointed to by the
+      /// provided `key`.
+      ///
+      /// `LocalKey`s are acquired using the `task_local` macro.
+      pub fn with_local<T, R, F>(&mut self, key: LocalKey<T>, f: F) -> R
+          where F: FnOnce(&mut T) -> R
+      {
+          ...
+      }
+  }
+
+  /// A handle used to awaken a task.
+  pub struct Waker {
+
+      /// Wakes up the associated task.
+      ///
+      /// This signals that the futures associated with the task are ready to
+      /// be `poll`ed again.
+      pub fn wake(&self) { ... }
+  }
+
+  ...
+}
+```
+
+The modified traits will look as follows:
+
+```rust
+pub trait Future {
+    type Item;
+    type Error;
```

Comment:
It was mentioned in a top level comment, but I prefer keeping discussion scoped to lines, so it's easier to follow specific discussions: **What about if `Future` simply had an `Item`, and no `Error`?** Combinators like `and_then` could exist for `F: Future<Item=<Result<T, E>>`.

Reply:
Good question! I've copy-pasted my response from below:

I can definitely see the advantages of such an approach, and I think the combinators could be adjusted appropriately. I could envision changing e.g. map and map_err into map, map_ok, and map_err or something similar. However, allowing non-Result futures also complicates many parts of the API and combinators-- now every combinator author has to standardize on whether to use Result-typed futures or "plain" futures. I anticipate that, in practice, most async operations involve some potentially fallible operation, so you'd have to manually specify Fut: IntoFuture<Item=Result<T, E>>, E: Into<Error> all over the place, which adds a lot of noise and extra type parameters to every function (T and E are now parameters instead of associated type projections). Infallible Futures are rare, and most cases can be represented reasonably via Future<Item = T, Error = !>.

Reply:
Could this be resolved with trait aliases?

    type Future<T, E> = CoreFuture<Item = Result<T, E>>;

Reply:
@rpjohnst Combinator methods such as `and_then` would need to be generic on `T` and `E` and have `where Self: Future<Result<Item = T, Error = E>>>` (or `where Self: Future<T, E>` with trait aliases). To me, the "result-y" future seems like it should be the default, and `CoreFuture` is an alias for `Future<Item = T, Error = !>`. Yes, it's slightly less ergonomic, but I think it's better to take the ergonomic hit on the non-result-y case than on the result-y case.

Reply:
That's why I named the alias `Future` rather than `ResultFuture`- no ergonomic hit on the result-y case, without putting `Error` in the trait.

Perhaps both the trait and the alias could be named `Future`, but in different modules, much like `io::Result` et al. Presumably there'd be no ergonomic hit in either case that way?

Reply:
To make things less abstract

### With `type Error`
```rust
fn and_then<F, B>(self, f: F) -> AndThen<Self, B, F>
        where F: FnOnce(Self::Item) -> B,
              B: IntoFuture<Error = Self::Error>,
              Self: Sized,
```

## Without `type Error`
```rust
fn and_then<T, E, U, F, B>(self, f: F) -> AndThen<Self, B, F>
        where F: FnOnce(T) -> B,
              B: IntoFuture<Item = Result<U, E>>,
              Self: Sized + Future<Item = Result<T, E>>,
```

Reply:
Without type Error looks fine. It is not like users implement combinators every day.

Having Error be part of future is akin to requiring all functions to return Result.

Reply:
> Having Error be part of future is akin to requiring all functions to return Result.

Is that a bad thing? Once you combine the first future with resulty behaviour, your whole chain is going to be resulty anyways. As previously stated, most futures implement fallible operations.

Futures are currently very easily explained as "future results".

Making and_then generic over 5 parameters looks incredibly noisy, IMHO.